### PR TITLE
Performance enhancements to validation and reference resolution

### PR DESCRIPTION
**Pull request by darrelmiller**
The primary problem resolved in this PR is the walker, which is used for both reference resolution and validation, was walking objects multiple times when they were $ref'd.  This has been resolved in the OpenApiWalker by adding a guard that does not walk the object if it has a non-null Reference and we are currently not walking through the Components section.

For the large 30MB, million line MS Graph OpenAPI, the time has gone from > 20mins, to ~ 10 secs on my Surface Pro 4.

I also added some statistics to the workbench to get an overview of the parsed file.  e.g. for the Graph API file it produces:

Path Items: 6783
Operations: 10609
Parameters: 23392
Request Bodies: 3815
Responses: 10609
Links: 132
Callbacks: 0
Schemas: 336995
  
Also updated Workbench to allow taking input from a file.

image

### CODE REVIEW

**Comment by PerthCharern** [comment]
Instead of creating a property InComponents, can we change the signature of Walk for Walk(OpenApiParameter parameter) to Walk(OpenApiParameter parameter, bool isInComponents) and pass the boolean value in the caller?

It seems a bit strange that OpenApiVisitorBase needs to keep a state.

File: src/Microsoft.OpenApi/Validations/OpenApiValidator.cs [hunk 1]
```diff
@@ -170,7 +169,15 @@ private void Validate<T>(T item)
         private void Validate(object item, Type type)
         {
             if (item == null) return;  // Required fields should be checked by higher level objects
-            var rules = _ruleSet.Where(r => r.ElementType == type);
+
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Validations/OpenApiValidator.cs]: nit, per style guideline and consistency, please convert this to 

if ( item == null ) 
{
    return;
}

File: src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs [hunk 2]
```diff
@@ -42,6 +56,12 @@ public static ValidationRuleSet GetDefaultRuleSet()
             return new ValidationRuleSet(_defaultRuleSet);
         }
 
+        public static ValidationRuleSet GetEmptyRuleSet()
+        {
+            // We create a new instance of ValidationRuleSet per call as a safeguard
+            // against unintentional modification of the private _defaultRuleSet.
+            return new ValidationRuleSet();
+        }
         /// <summary>
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs]: nit, newline

File: src/Microsoft.OpenApi.Readers/OpenApiStreamReader.cs [hunk 3]
```diff
@@ -90,11 +90,14 @@ public OpenApiDocument Read(Stream input, out OpenApiDiagnostic diagnostic)
             }
 
             // Validate the document
-            var errors = document.Validate(_settings.RuleSet);
-            foreach (var item in errors)
+            if (_settings.RuleSet != null && _settings.RuleSet.Rules.Count > 0)
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi.Readers/OpenApiStreamReader.cs]: >&& _settings.RuleSet.Rules.Count > 0 [](start = 42, length = 36)

just checking. Is this constraint needed?

File: src/Microsoft.OpenApi.Readers/Services/OpenApiReferenceResolver.cs [hunk 4]
```diff
@@ -176,7 +177,8 @@ private void ResolveObject<T>(T entity, Action<T> assign) where T : class, IOpen
 
             if (IsUnresolvedReference(entity))
             {
-                assign(ResolveReference<T>(entity.Reference));
+                var x = ResolveReference<T>(entity.Reference);
+                assign(x);
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi.Readers/Services/OpenApiReferenceResolver.cs]: Why this change?

File: test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
```
Comment (PerthCharern) [inline test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs]: nit, copyright

File: test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            };
+
+            // Act
+            var errors = document.Validate(new ValidationRuleSet() { new AllwaysFailRule<OpenApiSchema>()});
```
Comment (PerthCharern) [inline test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs]: >AllwaysFailRule [](start = 73, length = 15)

nit, Always typo

File: test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            // Act
+            var errors = document.Validate(new ValidationRuleSet() { new AllwaysFailRule<OpenApiSchema>() });
+
+            // Assert
+            Assert.True(errors.Count() == 0);
+        }
+
+    }
+
+    public class AllwaysFailRule<P> : ValidationRule<P> where P: IOpenApiElement
```
Comment (PerthCharern) [inline test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs]: >P [](start = 33, length = 1)

Not sure what P stands for. Maybe just use the standard T ?

File: test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+            OpenApiDocument document = new OpenApiDocument();
+            document.Components = new OpenApiComponents()
+            {
+                Schemas = new Dictionary<string, OpenApiSchema>()
+                {
+                    [sharedSchema.Reference.Id] = sharedSchema
+                }
+            };
+
```
Comment (PerthCharern) [inline test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs]: Should we also add this schema to an operation like above to test that scenario as well?

File: src/Microsoft.OpenApi.Workbench/MainWindow.xaml [hunk 9]
```diff
@@ -11,8 +11,24 @@
...
+        <Label>Input Content:</Label>
+      </StackPanel>
+      <TextBox x:Name="txtInput" Text="{Binding Input}" 
+               VerticalScrollBarVisibility="Auto" 
+               Margin="10" 
+               TextWrapping="Wrap" 
+               AcceptsReturn="True" 
+               FontFamily="Consolas"
+               VerticalContentAlignment="Stretch" />
+    </DockPanel>
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi.Workbench/MainWindow.xaml]: Do you mind posting a screenshot?

File: src/Microsoft.OpenApi.Workbench/StatsVisitor.cs [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.OpenApi.Models;
+using Microsoft.OpenApi.Services;
+
+namespace Microsoft.OpenApi.Workbench
+{
+    public class StatsVisitor : OpenApiVisitorBase
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi.Workbench/StatsVisitor.cs]: >public [](start = 4, length = 6)

nit, document public class/properties

**Comment by darrelmiller** [comment]
I considered passing it as a parameter, but as Components can hold many different types, it would require passing that value almost everywhere.  Also, we already hold state while walking.  We hold the Path in VisitorBase and we hold stacks for tracking loops in the the Walker.  It does make me wonder why we have state in both.  Maybe the state can be moved into the walker instead of being in the Visitor.

---
In reply to: [382076651](https://github.com/Microsoft/OpenAPI.NET/pull/241#issuecomment-382076651) [](ancestors = 382076651)

File: src/Microsoft.OpenApi.Readers/OpenApiStreamReader.cs (hunk 3 above)
Comment (darrelmiller) [reply src/Microsoft.OpenApi.Readers/OpenApiStreamReader.cs]: It's there for performance.  Otherwise we walk the entire graph and do no validation.

---
In reply to: [182166250](https://github.com/Microsoft/OpenAPI.NET/pull/241#discussion_r182166250) [](ancestors = 182166250)

File: src/Microsoft.OpenApi.Readers/Services/OpenApiReferenceResolver.cs (hunk 4 above)
Comment (darrelmiller) [reply src/Microsoft.OpenApi.Readers/Services/OpenApiReferenceResolver.cs]: I'll revert.  I did it temporarily whilst hunting down what was taking so long.

---
In reply to: [182166389](https://github.com/Microsoft/OpenAPI.NET/pull/241#discussion_r182166389) [](ancestors = 182166389)

**Comment by darrelmiller** [comment]
I was able to move the InComponents state into the walker.  We still have the Path Stack in the visitor and the ValidationVisitor holds the list of errors detected.  So, I don't think we can make the Visitor completely stateless, but we can at least keep it to a minimum.

---
In reply to: [382087733](https://github.com/Microsoft/OpenAPI.NET/pull/241#issuecomment-382087733) [](ancestors = 382087733,382076651)

File: src/Microsoft.OpenApi/Services/OpenApiWalker.cs [hunk 11]
```diff
@@ -37,6 +38,9 @@ public void Walk(OpenApiDocument doc)
             {
                 return;
             }
+            _schemaLoop.Clear();
```
Comment (PerthCharern) [inline src/Microsoft.OpenApi/Services/OpenApiWalker.cs]: nit, newline

**Comment by PerthCharern** [comment]
Sure. This looks okay to me.

---
In reply to: [384155854](https://github.com/Microsoft/OpenAPI.NET/pull/241#issuecomment-384155854) [](ancestors = 384155854,382087733,382076651)

File: test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+            // Act
+            var errors = document.Validate(new ValidationRuleSet() { new AlwaysFailRule<OpenApiSchema>()});
+
+
+            // Assert
+            Assert.True(errors.Count() == 1);            
+        }
+
+        [Fact]
+        public void UnResolvedReferencedSchemaShouldNotBeValidated()
```
Comment (PerthCharern) [inline test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs]: >UnResolvedReferencedSchemaShouldNotBeValidated [](start = 20, length = 46)

nit, casing of unresolved incorrect:

UnresolvedReferencedSchemaShouldNotBeValidated

File: test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+
+            // Act
+            var errors = document.Validate(new ValidationRuleSet() { new AlwaysFailRule<OpenApiSchema>() });
+
+            // Assert
+            Assert.True(errors.Count() == 0);
+        }
+
+        [Fact]
+        public void UnResolvedSchemaReferencedShouldNotBeValidated()
```
Comment (PerthCharern) [inline test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs]: >UnResolvedSchemaReferencedShouldNotBeValidated [](start = 20, length = 46)

UnresolvedReferenceSchemaInOperationShouldNotBeValidated

File: src/Microsoft.OpenApi.Workbench/MainWindow.xaml (hunk 9 above)
Comment (PerthCharern) [reply src/Microsoft.OpenApi.Workbench/MainWindow.xaml]: I still don't see the screenshot, but I think it's alright.

---
In reply to: [182167917](https://github.com/Microsoft/OpenAPI.NET/pull/241#discussion_r182167917) [](ancestors = 182167917)

File: test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs (hunk 12 above)
Comment (darrelmiller) [reply test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs]: Fixed the casing issue.  One tests is ReferencedSchema, the other is SchemaReference

---
In reply to: [184120408](https://github.com/Microsoft/OpenAPI.NET/pull/241#discussion_r184120408) [](ancestors = 184120408)

Comment (PerthCharern) [reply test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs]: Ah, ok. That makes sense. The other one has a typo though ("Referenced" instead of "Reference")

---
In reply to: [184126122](https://github.com/Microsoft/OpenAPI.NET/pull/241#discussion_r184126122) [](ancestors = 184126122,184120408)

File: src/Microsoft.OpenApi.Workbench/StatsVisitor.cs [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.OpenApi.Models;
+using Microsoft.OpenApi.Services;
+
+namespace Microsoft.OpenApi.Workbench
+{
+    internal class StatsVisitor : OpenApiVisitorBase
```
[the remark this answers is not in the document]
Comment (darrelmiller) [reply src/Microsoft.OpenApi.Workbench/StatsVisitor.cs]: I put it here as a quick and dirty experimental statistics tool.  I don't want to commit to something in the core library until we actually spend a bit of time designing it.

---
In reply to: [184122675](https://github.com/Microsoft/OpenAPI.NET/pull/241#discussion_r184122675) [](ancestors = 184122675)

**Comment by darrelmiller** [comment]
>         internal void Validate()

I renamed it ParseDocument.  It's called Convert on the UI but I'm not a fan of that either.

---
In reply to: [384346044](https://github.com/Microsoft/OpenAPI.NET/pull/241#issuecomment-384346044) [](ancestors = 384346044)

---
Refers to: src/Microsoft.OpenApi.Workbench/MainModel.cs:166 in 7e51f03. [](commit_id = 7e51f0377d793246ea3b2b6a2fad31c11a3d6095, deletion_comment = False)

File: test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs (hunk 12 above)
Comment (darrelmiller) [reply test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs]: "Referenced" was intentional, but Reference works too.  Changed it.

---
In reply to: [184126692](https://github.com/Microsoft/OpenAPI.NET/pull/241#discussion_r184126692) [](ancestors = 184126692,184126122,184120408)

File: src/Microsoft.OpenApi.Workbench/MainWindow.xaml (hunk 9 above)
Comment (darrelmiller) [reply src/Microsoft.OpenApi.Workbench/MainWindow.xaml]: I put it in the PR description.  I wasn't sure where else to put it.

---
In reply to: [184121152](https://github.com/Microsoft/OpenAPI.NET/pull/241#discussion_r184121152) [](ancestors = 184121152,182167917)

File: test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+            // Act
+            var errors = document.Validate(new ValidationRuleSet() { new AlwaysFailRule<OpenApiSchema>()});
+
+
+            // Assert
+            Assert.True(errors.Count() == 1);            
+        }
+
+        [Fact]
+        public void UnresolvedReferenceSchemaShouldNotBeValidated()
```
Comment (PerthCharern) [inline test/Microsoft.OpenApi.Tests/Validations/OpenApiReferenceValidationTests.cs]: >UnresolvedReferenceSchemaShouldNotBeValidated [](start = 20, length = 45)

I was thinking of the other way round. "Referenced" here and "Reference" below. Sorry I wasn't clear :(

File: src/Microsoft.OpenApi.Workbench/MainWindow.xaml (hunk 9 above)
Comment (PerthCharern) [reply src/Microsoft.OpenApi.Workbench/MainWindow.xaml]: Saw it now. Thanks!

---
In reply to: [184128635](https://github.com/Microsoft/OpenAPI.NET/pull/241#discussion_r184128635) [](ancestors = 184128635,184121152,182167917)
